src/Rules/NoUnnecessaryCollectionCallRule.php

```diff
@@ -0,0 +1,331 @@
+            if (count($node->args) === 0) {
+                // 'first', also accepts a closure as an argument.
+                return [$this->formatError($name->toString())];
+            }
+        } elseif ($this->isRiskyMethod($name)) {
+            return [$this->formatError($name->toString())];
+        } elseif ($this->isRiskyParamMethod($name)) {
+            if (count($node->args) === 0) {
+                // Calling e.g. DB::table()->pluck('age')-sum()
+                return [$this->formatError($name->toString())];
+            }
+            if ($this->firstArgIsDatabaseColumn($node, $scope)) {
+                return [$this->formatError($name->toString())];
+            }
+        } elseif (in_array($name->toLowerString(), ['contains', 'containsstrict'], true)) {
+            // 'contains' can also be called with Model instances or keys as its first argument
+            /** @var \PhpParser\Node\Arg[] $args */
+            $args = $node->args;
+            if (count($args) === 1 && ! ($args[0]->value instanceof Node\Expr\Closure)) {
+                return [$this->formatError($name->toString())];
+            }
+
+            if ($this->firstArgIsDatabaseColumn($node, $scope)) {
+                return [$this->formatError($name->toString())];
+            }
+        }
+
+        return [];
+    }
+
+    /**
+     * Determines whether the first argument is a string and references a database column.
+     * @param MethodCall $node
+     * @param Scope $scope
+     * @return bool
+     */
+    protected function firstArgIsDatabaseColumn(MethodCall $node, Scope $scope): bool
+    {
+        /** @var \PhpParser\Node\Arg[] $args */
+        $args = $node->args;
+        if (count($args) === 0 || ! ($args[0]->value instanceof Node\Scalar\String_)) {
+            return false;
+        }
+
+        $iterableType = $scope->getType($node->var)->getIterableValueType();
+        if ($iterableType instanceof MixedType) {
+            $previous_call = $node->var;
+            if ($previous_call instanceof MethodCall) {
+                $query_builder_type = $scope->getType($previous_call->var);
+                if ((new ObjectType(QueryBuilder::class))->isSuperTypeOf($query_builder_type)->yes()) {
+                    // We encountered a DB query such as DB::table(..)->get()->max('id')
+                    // We assume max('id') could have been retrieved without calling get().
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /** @var \PHPStan\Type\ObjectType $iterableType */
+        if ((new ObjectType(Model::class))->isSuperTypeOf($iterableType)->yes()) {
+            $modelReflection = $this->reflectionProvider->getClass($iterableType->getClassName());
+
+            /** @var \PhpParser\Node\Scalar\String_ $firstArg */
+            $firstArg = $args[0]->value;
+
+            return $this->propertyExtension->hasProperty($modelReflection, $firstArg->value);
```

Comment:
`ModelPropertyExtension` only works if we can read the database migrations. Also this will ignore if the property is defined by annotations on the class.

So I think it's better to use `return $modelReflection->hasProperty($firstArg->value);` here. And we can also remove the `ModelPropertyExtension` injection in the constructor.

Reply:
I chose to use the ModelPropertyExtension, because as far as I understood this will tell us whether the property exists as a column in our database. 

Using `return $modelReflection->hasProperty($firstArg->value);` will result in false positives because the `hasProperty` on the `modelReflection` will return `true` for, for example, relations. 

E.g. the following should be valid:
```
User::with(['accounts'])->get()->pluck('accounts');
```

but if we use the modelReflection it will be marked as invalid.

> Also this will ignore if the property is defined by annotations on the class.

That's what we want, right? If there is a PHPDoc property on a Model, we have no idea where that property comes from. We can not assume that the property exists as a column in our database because that will lead to falsely reporting errors again.

Reply:
Hmm. Yeah, that's a good point! So it's good like this.

Reply:
I totally understand your reasoning, although I personally don't think we should rely on database migration files for this. Especially since in laravel 8, migrations are being optimized via dump commands and we won't be able to consistently rely on them moving forward. 

Although  I am biased as I use a more scalable solution [Skeema](https://github.com/skeema/skeema) for managing migrations

Reply:
It's my understanding that since this PR is merged, this package explicitly depends on database migrations being present, or at least when this rule is enabled.

Reply:
If you don't have migrations, the `ModelPropertyExtension` would return false for `hasProperty`, therefore the rule will not give you an error. It gives an error when we are really sure the given string is a database column.

Reply:
wohoo, i inverted that check in my head when going through the logic